src/ExpressionEvaluator/Core/Source/ExpressionCompiler/DynamicFlagsCustomTypeInfo.cs

```diff
@@ -13,25 +12,55 @@ internal struct DynamicFlagsCustomTypeInfo
         /// <remarks>Internal for testing.</remarks>
         internal static readonly Guid PayloadTypeId = new Guid("826D6EC1-DC4B-46AF-BE05-CD3F1A1FD4AC");
 
-        private readonly BitArray _bits;
+        private readonly ReadOnlyCollection<byte> _bytes;
 
-        public DynamicFlagsCustomTypeInfo(BitArray bits)
+        private DynamicFlagsCustomTypeInfo(ReadOnlyCollection<byte> bytes)
+        {
+            _bytes = bytes;
+        }
+
+        public DynamicFlagsCustomTypeInfo(Guid payloadTypeId, ReadOnlyCollection<byte> payload)
+            : this(payloadTypeId == PayloadTypeId ? payload : null)
         {
-            _bits = bits;
         }
 
         public DynamicFlagsCustomTypeInfo(DkmClrCustomTypeInfo typeInfo)
+            : this(typeInfo != null && typeInfo.PayloadTypeId == PayloadTypeId ? typeInfo.Payload : null)
+        {
+        }
+
+        internal DynamicFlagsCustomTypeInfo(bool[] dynamicFlags)
```

Comment:
Would it help to change the signature of this .ctor to: ArrayBuilder<bool> dynamicFlags, int startIndex)? That would avoid copying DynamicTransformsEncoder.Encode() results to a bool[] from the heap and it would allow SkipOne() to be expressed as: this.CopyTo(builder); result = new DynamicFlagsCustomTypeInfo(builder, 1);.